File: utils/formatters.py
def format_vnd(amount) -> str:
    if amount is None or amount != amount:
        return "N/A"
    amount = float(amount)
    if abs(amount) >= 1_000_000_000:
        return f"{amount / 1_000_000_000:,.1f}B ₫"
    if abs(amount) >= 1_000_000:
        return f"{amount / 1_000_000:,.0f}M ₫"
    if abs(amount) >= 1_000:
        return f"{amount / 1_000:,.0f}K ₫"
    return f"{amount:,.0f} ₫"


def format_number(n) -> str:
    if n is None or n != n:
        return "N/A"
    n = float(n)
    if abs(n) >= 1_000_000:
        return f"{n / 1_000_000:,.1f}M"
    if abs(n) >= 1_000:
        return f"{n / 1_000:,.1f}K"
    return f"{n:,.0f}"


def format_pct(value, decimals=1) -> str:
    if value is None or value != value:
        return "N/A"
    return f"{value:.{decimals}f}%"

File: frontend/src/lib/utils/formatters.ts
export function formatVND(amount: number | null | undefined): string {
  if (amount == null || isNaN(amount)) return 'N/A'
  if (Math.abs(amount) >= 1_000_000_000) {
    return `${(amount / 1_000_000_000).toFixed(1)}B ₫`
  }
  if (Math.abs(amount) >= 1_000_000) {
    return `${(amount / 1_000_000).toFixed(1)}M ₫`
  }
  return new Intl.NumberFormat('vi-VN', { style: 'currency', currency: 'VND' }).format(amount)
}

export function formatNumber(n: number | null | undefined): string {
  if (n == null || isNaN(n)) return 'N/A'
  if (Math.abs(n) >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`
  if (Math.abs(n) >= 1_000) return `${(n / 1_000).toFixed(1)}K`
  return n.toLocaleString()
}

export function formatPct(value: number | null | undefined, decimals = 1): string {
  if (value == null || isNaN(value)) return 'N/A'
  return `${value.toFixed(decimals)}%`
}

export function formatUSD(amount: number | null | undefined): string {
  if (amount == null || isNaN(amount)) return 'N/A'
  if (Math.abs(amount) >= 1_000) return `$${(amount / 1_000).toFixed(0)}K`
  return `$${amount.toLocaleString()}`
}
